File: conformance/encode-json-utf8-lf-v0/encoder.py
#!/usr/bin/env python3
"""Independent Python encoder for encode-json-utf8-lf.v0.

The object serializer is local to this contract. Only the scalar binary64
rendering rule is aligned with RFC 8785 section 3.2.2.3 / ECMAScript. Python's
json.dumps float rendering and Python's native code-point key ordering are not
used for normative output.
"""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
import math
import pathlib
import struct
import sys
from typing import Any


SAFE_INTEGER = 9007199254740991
KNOWN_ERRORS = {
    "INTEGER_OUT_OF_RANGE",
    "NEGATIVE_ZERO",
    "NON_FINITE_NUMBER",
    "NON_SCALAR_KEY",
    "NON_SCALAR_STRING",
    "NUMBER_NOT_EXACTLY_BINARY64",
}


class DomainError(ValueError):
    def __init__(self, category: str):
        if category not in KNOWN_ERRORS:
            raise ValueError(f"unknown domain-error category: {category}")
        super().__init__(category)
        self.category = category


@dataclass(frozen=True)
class F64Value:
    bits: int
    value: float


@dataclass(frozen=True)
class IntegerValue:
    value: int


@dataclass(frozen=True)
class ObjectValue:
    entries: tuple[tuple[str, Any], ...]


def _is_scalar_string(value: str) -> bool:
    return all(not 0xD800 <= ord(ch) <= 0xDFFF for ch in value)


def _utf16_units(value: str) -> tuple[int, ...]:
    if not _is_scalar_string(value):
        raise DomainError("NON_SCALAR_KEY")
    raw = value.encode("utf-16-be")
    return tuple(int.from_bytes(raw[i : i + 2], "big") for i in range(0, len(raw), 2))


def _escape_string(value: str, *, key: bool = False, allow_surrogate: bool = False) -> str:
    out = ['"']
    short = {8: "\\b", 9: "\\t", 10: "\\n", 12: "\\f", 13: "\\r"}
    for ch in value:
        cp = ord(ch)
        if 0xD800 <= cp <= 0xDFFF:
            if allow_surrogate:
                out.append(f"\\u{cp:04x}")
                continue
            raise DomainError("NON_SCALAR_KEY" if key else "NON_SCALAR_STRING")
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif cp in short:
            out.append(short[cp])
        elif cp <= 0x1F:
            out.append(f"\\u{cp:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _ecmascript_number(value: float) -> str:
    """Render one already-validated binary64 value using ECMAScript spelling.

    CPython's repr supplies the shortest round-tripping coefficient. This code
    independently applies ECMAScript's fixed/exponent thresholds and exponent
    syntax; json.dumps is deliberately not involved.
    """
    if value == 0.0:
        return "0"
    sign = "-" if value < 0 else ""
    text = repr(abs(value)).lower()
    exponent = 0
    if "e" in text:
        mantissa, exponent_text = text.split("e", 1)
        exponent = int(exponent_text)
    else:
        mantissa = text
    if "." in mantissa:
        first, last = mantissa.split(".", 1)
    else:
        first, last = mantissa, ""
    if last == "0":
        last = ""
    if "e" not in text:
        return sign + first + ("." + last if last else "")
    if 0 < exponent < 21:
        digits = first + last
        zero_count = exponent + 1 - len(digits)
        if zero_count >= 0:
            return sign + digits + ("0" * zero_count)
        split = exponent + 1
        return sign + digits[:split] + "." + digits[split:]
    if -7 < exponent < 0:
        return sign + "0." + ("0" * (-exponent - 1)) + first + last
    exponent_sign = "+" if exponent >= 0 else "-"
    return sign + first + ("." + last if last else "") + "e" + exponent_sign + str(abs(exponent))


def _render_number(value: F64Value | IntegerValue, fault: str | None) -> str:
    permit_unsafe = fault == "permit_unsafe_integer"
    if isinstance(value, IntegerValue):
        if not permit_unsafe and abs(value.value) > SAFE_INTEGER:
            raise DomainError("INTEGER_OUT_OF_RANGE")
        number = float(value.value)
        if int(number) != value.value:
            raise DomainError("NUMBER_NOT_EXACTLY_BINARY64")
    else:
        number = value.value
        if not math.isfinite(number):
            raise DomainError("NON_FINITE_NUMBER")
        if number == 0.0 and value.bits >> 63:
            if fault != "permit_negative_zero":
                raise DomainError("NEGATIVE_ZERO")
        if number.is_integer() and not permit_unsafe and abs(number) > SAFE_INTEGER:
            raise DomainError("INTEGER_OUT_OF_RANGE")
    if fault == "python_native_numbers":
        return repr(number)
    return _ecmascript_number(number)


def _serialize(value: Any, fault: str | None = None) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (F64Value, IntegerValue)):
        return _render_number(value, fault)
    if isinstance(value, str):
        return _escape_string(value, allow_surrogate=fault == "permit_lone_surrogate")
    if isinstance(value, list):
        return "[" + ",".join(_serialize(item, fault) for item in value) + "]"
    if isinstance(value, ObjectValue):
        seen: set[str] = set()
        for key, _ in value.entries:
            if key in seen:
                raise ValueError("fixture transport contains duplicate object key")
            seen.add(key)
            if not _is_scalar_string(key) and fault != "permit_lone_surrogate":
                raise DomainError("NON_SCALAR_KEY")
        if fault == "codepoint_sort":
            entries = sorted(value.entries, key=lambda entry: entry[0])
        elif fault == "permit_lone_surrogate":
            entries = sorted(value.entries, key=lambda entry: tuple(ord(c) for c in entry[0]))
        else:
            entries = sorted(value.entries, key=lambda entry: _utf16_units(entry[0]))
        return "{" + ",".join(
            _escape_string(key, key=True, allow_surrogate=fault == "permit_lone_surrogate")
            + ":"
            + _serialize(item, fault)
            for key, item in entries
        ) + "}"
    raise TypeError(f"unsupported abstract value: {type(value).__name__}")


def encode(value: Any, fault: str | None = None) -> bytes:
    body = _serialize(value, fault).encode("utf-8")
    return body if fault == "missing_terminal_lf" else body + b"\n"


def decode_transport(node: dict[str, Any]) -> Any:
    kind = node.get("type")
    if kind == "null":
        return None
    if kind == "boolean":
        return bool(node["value"])
    if kind == "string":
        return node["value"]
    if kind == "integer":
        decimal = node["decimal"]
        if not isinstance(decimal, str) or not decimal or decimal in {"+", "-"}:
            raise ValueError("invalid integer carrier")
        return IntegerValue(int(decimal, 10))
    if kind == "f64_bits":
        raw_hex = node["hex"]
        if not isinstance(raw_hex, str) or len(raw_hex) != 16:
            raise ValueError("invalid f64_bits carrier")
        raw = bytes.fromhex(raw_hex)
        return F64Value(int.from_bytes(raw, "big"), struct.unpack(">d", raw)[0])
    if kind == "array":
        return [decode_transport(item) for item in node["items"]]
    if kind == "object":
        return ObjectValue(tuple((entry["key"], decode_transport(entry["value"])) for entry in node["entries"]))
    raise ValueError(f"unknown fixture transport type: {kind!r}")


def run_vectors(document: dict[str, Any], fault: str | None = None) -> dict[str, dict[str, Any]]:
    results: dict[str, dict[str, Any]] = {}
    for vector in document["vectors"]:
        vector_id = vector["id"]
        try:
            encoded = encode(decode_transport(vector["input"]), fault)
            results[vector_id] = {
                "status": "success",
                "bytes_hex": encoded.hex(),
                "byte_length": len(encoded),
                "sha256": hashlib.sha256(encoded).hexdigest(),
            }
        except DomainError as error:
            results[vector_id] = {"status": "rejection", "error": error.category}
    return results


def _self_grade(document: dict[str, Any], results: dict[str, dict[str, Any]]) -> list[str]:
    failures = []
    expected_ids = {vector["id"] for vector in document["vectors"]}
    if set(results) != expected_ids:
        failures.append("result id set does not match vector id set")
    for vector in document["vectors"]:
        if results.get(vector["id"]) != vector.get("expect"):
            failures.append(vector["id"])
    return failures


def main() -> int:
    grade = "--grade" in sys.argv
    fault = None
    if "--negative-control" in sys.argv:
        index = sys.argv.index("--negative-control")
        fault = sys.argv[index + 1]
    if grade:
        document = json.load(sys.stdin)
    else:
        path = pathlib.Path(__file__).with_name("encode-json-utf8-lf-v0.vectors.json")
        document = json.loads(path.read_text(encoding="utf-8"))
    results = run_vectors(document, fault)
    if grade:
        print(json.dumps(results, ensure_ascii=True, separators=(",", ":")))
        return 0
    failures = _self_grade(document, results)
    successes = sum(result["status"] == "success" for result in results.values())
    rejections = len(results) - successes
    print(f"Python adapter: {len(results) - len(failures)}/{len(results)} vectors; {successes} success; {rejections} rejection")
    if failures:
        for vector_id in failures:
            print(f"FAIL {vector_id}")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: conformance/encode-json-utf8-lf-v0/encoder.ts
#!/usr/bin/env bun
/** Independent TypeScript encoder for encode-json-utf8-lf.v0. */

import { createHash } from "node:crypto";

const SAFE_INTEGER = 9007199254740991;
const KNOWN_ERRORS = new Set([
  "INTEGER_OUT_OF_RANGE",
  "NEGATIVE_ZERO",
  "NON_FINITE_NUMBER",
  "NON_SCALAR_KEY",
  "NON_SCALAR_STRING",
  "NUMBER_NOT_EXACTLY_BINARY64",
]);

class DomainError extends Error {
  category: string;

  constructor(category: string) {
    if (!KNOWN_ERRORS.has(category)) throw new Error(`unknown domain-error category: ${category}`);
    super(category);
    this.category = category;
  }
}

type F64Value = { kind: "f64"; bits: bigint; value: number };
type IntegerValue = { kind: "integer"; value: bigint };
type ObjectValue = { kind: "object"; entries: Array<[string, AbstractValue]> };
type AbstractValue = null | boolean | string | F64Value | IntegerValue | ObjectValue | AbstractValue[];

type Transport = {
  type: string;
  value?: unknown;
  decimal?: string;
  hex?: string;
  items?: Transport[];
  entries?: Array<{ key: string; value: Transport }>;
};

type VectorDocument = {
  vectors: Array<{ id: string; input: Transport; expect?: AdapterResult }>;
};

type AdapterResult =
  | { status: "success"; bytes_hex: string; byte_length: number; sha256: string }
  | { status: "rejection"; error: string };

function scalarError(value: string, key: boolean): string | null {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      if (index + 1 >= value.length) return key ? "NON_SCALAR_KEY" : "NON_SCALAR_STRING";
      const low = value.charCodeAt(index + 1);
      if (low < 0xdc00 || low > 0xdfff) return key ? "NON_SCALAR_KEY" : "NON_SCALAR_STRING";
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return key ? "NON_SCALAR_KEY" : "NON_SCALAR_STRING";
    }
  }
  return null;
}

function compareUtf16(left: string, right: string): number {
  const shared = Math.min(left.length, right.length);
  for (let index = 0; index < shared; index += 1) {
    const a = left.charCodeAt(index);
    const b = right.charCodeAt(index);
    if (a !== b) return a < b ? -1 : 1;
  }
  return left.length - right.length;
}

function escapeString(value: string, key = false): string {
  const invalid = scalarError(value, key);
  if (invalid) throw new DomainError(invalid);
  let out = '"';
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      out += value[index] + value[index + 1];
      index += 1;
      continue;
    }
    if (unit === 0x22) out += '\\"';
    else if (unit === 0x5c) out += "\\\\";
    else if (unit === 0x08) out += "\\b";
    else if (unit === 0x0c) out += "\\f";
    else if (unit === 0x0a) out += "\\n";
    else if (unit === 0x0d) out += "\\r";
    else if (unit === 0x09) out += "\\t";
    else if (unit <= 0x1f) out += `\\u${unit.toString(16).padStart(4, "0")}`;
    else out += value[index];
  }
  return out + '"';
}

function renderNumber(value: F64Value | IntegerValue): string {
  let number: number;
  if (value.kind === "integer") {
    const bound = 9007199254740991n;
    if (value.value < -bound || value.value > bound) throw new DomainError("INTEGER_OUT_OF_RANGE");
    number = Number(value.value);
    if (BigInt(number) !== value.value) throw new DomainError("NUMBER_NOT_EXACTLY_BINARY64");
  } else {
    number = value.value;
    if (!Number.isFinite(number)) throw new DomainError("NON_FINITE_NUMBER");
    if (Object.is(number, -0)) throw new DomainError("NEGATIVE_ZERO");
    if (Number.isInteger(number) && Math.abs(number) > SAFE_INTEGER) {
      throw new DomainError("INTEGER_OUT_OF_RANGE");
    }
  }
  // This call occurs only after the v0 domain has accepted the scalar. It is
  // the isolated ECMAScript/RFC 8785 section 3.2.2.3 number-rendering step.
  return JSON.stringify(number);
}

function serialize(value: AbstractValue): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return escapeString(value);
  if (Array.isArray(value)) return `[${value.map(serialize).join(",")}]`;
  if (value.kind === "f64" || value.kind === "integer") return renderNumber(value);
  const seen = new Set<string>();
  for (const [key] of value.entries) {
    const invalid = scalarError(key, true);
    if (invalid) throw new DomainError(invalid);
    if (seen.has(key)) throw new Error("fixture transport contains duplicate object key");
    seen.add(key);
  }
  const entries = [...value.entries].sort((a, b) => compareUtf16(a[0], b[0]));
  return `{${entries.map(([key, item]) => `${escapeString(key, true)}:${serialize(item)}`).join(",")}}`;
}

function encode(value: AbstractValue): Uint8Array {
  return new TextEncoder().encode(`${serialize(value)}\n`);
}

function decodeTransport(node: Transport): AbstractValue {
  if (node.type === "null") return null;
  if (node.type === "boolean") return Boolean(node.value);
  if (node.type === "string") return String(node.value);
  if (node.type === "integer") {
    if (typeof node.decimal !== "string" || !/^-?(0|[1-9][0-9]*)$/.test(node.decimal)) {
      throw new Error("invalid integer carrier");
    }
    return { kind: "integer", value: BigInt(node.decimal) };
  }
  if (node.type === "f64_bits") {
    if (typeof node.hex !== "string" || !/^[0-9a-f]{16}$/.test(node.hex)) {
      throw new Error("invalid f64_bits carrier");
    }
    const bits = BigInt(`0x${node.hex}`);
    const buffer = new ArrayBuffer(8);
    const view = new DataView(buffer);
    view.setBigUint64(0, bits, false);
    return { kind: "f64", bits, value: view.getFloat64(0, false) };
  }
  if (node.type === "array") return (node.items ?? []).map(decodeTransport);
  if (node.type === "object") {
    return {
      kind: "object",
      entries: (node.entries ?? []).map((entry) => [entry.key, decodeTransport(entry.value)]),
    };
  }
  throw new Error(`unknown fixture transport type: ${node.type}`);
}

function runVectors(document: VectorDocument): Record<string, AdapterResult> {
  const results: Record<string, AdapterResult> = {};
  for (const vector of document.vectors) {
    try {
      const bytes = encode(decodeTransport(vector.input));
      results[vector.id] = {
        status: "success",
        bytes_hex: Buffer.from(bytes).toString("hex"),
        byte_length: bytes.byteLength,
        sha256: createHash("sha256").update(bytes).digest("hex"),
      };
    } catch (error) {
      if (!(error instanceof DomainError)) throw error;
      results[vector.id] = { status: "rejection", error: error.category };
    }
  }
  return results;
}

function selfGrade(document: VectorDocument, results: Record<string, AdapterResult>): string[] {
  const expectedIds = document.vectors.map((vector) => vector.id).sort();
  const actualIds = Object.keys(results).sort();
  const failures: string[] = [];
  if (JSON.stringify(expectedIds) !== JSON.stringify(actualIds)) failures.push("result id set");
  for (const vector of document.vectors) {
    if (JSON.stringify(results[vector.id]) !== JSON.stringify(vector.expect)) failures.push(vector.id);
  }
  return failures;
}

async function main(): Promise<number> {
  const grade = Bun.argv.includes("--grade");
  const document = grade
    ? (JSON.parse(await Bun.stdin.text()) as VectorDocument)
    : (JSON.parse(await Bun.file(`${import.meta.dir}/encode-json-utf8-lf-v0.vectors.json`).text()) as VectorDocument);
  const results = runVectors(document);
  if (grade) {
    process.stdout.write(`${JSON.stringify(results)}\n`);
    return 0;
  }
  const failures = selfGrade(document, results);
  const successes = Object.values(results).filter((result) => result.status === "success").length;
  console.log(`TypeScript adapter: ${document.vectors.length - failures.length}/${document.vectors.length} vectors; ${successes} success; ${document.vectors.length - successes} rejection`);
  for (const vectorId of failures) console.log(`FAIL ${vectorId}`);
  return failures.length ? 1 : 0;
}

if (import.meta.main) process.exit(await main());

export { decodeTransport, encode, runVectors };
